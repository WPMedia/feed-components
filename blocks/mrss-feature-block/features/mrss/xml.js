import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import get from 'lodash/get'
import moment from 'moment'
import getProperties from 'fusion:properties'
import { resizerKey } from 'fusion:environment'
import { BuildContent } from '@wpmedia/feeds-content-elements'
import { generatePropsForFeed } from '@wpmedia/feeds-prop-types'
import { buildResizerURL } from '@wpmedia/feeds-resizer'
import { findVideo } from '@wpmedia/feeds-find-video-stream'

const jmespath = require('jmespath')

const rssTemplate = (
  elements,
  {
    channelTitle,
    channelDescription,
    channelPath,
    channelCopyright,
    channelTTL,
    channelUpdatePeriod,
    channelUpdateFrequency,
    channelCategory,
    channelLogo,
    imageTitle,
    imageCaption,
    imageCredits,
    itemTitle,
    itemDescription,
    pubDate,
    itemCredits,
    itemCategory,
    includeContent,
    resizerURL,
    domain,
    feedTitle,
    feedLanguage,
    videoInfo,
    mrssBuildContent,
  },
) => ({
  rss: {
    '@xmlns:atom': 'http://www.w3.org/2005/Atom',
    '@version': '2.0',
    '@xmlns:media': 'http://search.yahoo.com/mrss/',
    channel: {
      title: `${channelTitle || feedTitle + ' Videos'}`,
      link: domain,
      'atom:link': {
        '@href': `${domain}${channelPath}`,
        '@rel': 'self',
        '@type': 'application/rss+xml',
      },
      description: channelDescription || feedTitle,
      pubDate: moment.utc(Date.now()).format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
      ...(channelLogo && {
        image: {
          url: buildResizerURL(channelLogo, resizerKey, resizerURL),
          title: `${channelTitle || feedTitle + ' Videos'}`,
          link: domain,
        },
      }),

      item: elements.map((s) => {
        const url = `${domain}${s.website_url || s.canonical_url}`
        const img =
          s.promo_items && (s.promo_items.basic || s.promo_items.lead_art)
        const videoStream = findVideo(s, videoInfo)

        return {
          title: `${jmespath.search(s, itemTitle)}`,
          link: url,
          ...(itemCredits &&
            (jmespath.search(img, itemCredits) || []).length && {
              'media:credit': {
                '@role': 'producer',
                $: jmespath.search(img, itemCredits).join(','),
              },
            }),
          ...(itemDescription && {
            description: jmespath.search(s, itemDescription),
          }),
          guid: {
            '@isPermaLink': false,
            '#': url,
          },
          referenceid: s._id,
          pubDate: moment
            .utc(s[pubDate])
            .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),

          'media:content': {
            '@': {
              isDefault: 'true',
              ...(s.duration && {
                duration: Math.trunc(s.duration / 1000),
              }),
              ...(videoStream && {
                ...(videoInfo.url && { url: videoInfo.url }),
                ...(videoInfo.height && { height: videoInfo.height }),
                ...(videoInfo.width && { width: videoInfo.width }),
                ...(videoInfo.bitrate && { bitrate: videoInfo.bitrate }),
                ...(videoInfo.stream_type && {
                  type: videoInfo.stream_type,
                  '#': videoStream,
                }),
              }),
            },
          },
          'media:keywords': (
            jmespath.search(s, 'taxonomy.seo_keywords[*]') || []
          ).join(','),
          ...(s.description ||
            (s.subheadlines &&
              s.subheadlines.basic && {
                'media:caption': {
                  $: s.description || s.subheadlines.basic,
                },
              })),
          'media:transcript': s.transcript,
          ...(s.taxonomy &&
            s.taxonomy.primary_section &&
            s.taxonomy.primary_section.name && {
              'media:category': s.taxonomy.primary_section.name,
            }),
          ...(img &&
            img.url && {
              'media:thumbnail': {
                '@url': buildResizerURL(img.url, resizerKey, resizerURL),
              },
            }),
        }
      }),
    },
  },
})

export function Mrss({ globalContent, customFields, arcSite }) {
  const {
    resizerURL = '',
    feedDomainURL = '',
    feedTitle = '',
    feedLanguage = '',
  } = getProperties(arcSite)

  function MrssBuildContent() {
    BuildContent.call(this)
  }
  const mrssBuildContent = new MrssBuildContent()

  // can't return null for xml return type, must return valid xml template
  return rssTemplate(get(globalContent, 'content_elements', []), {
    ...customFields,
    resizerURL,
    domain: feedDomainURL,
    feedTitle,
    feedLanguage,
    mrssBuildContent,
  })
}

Mrss.propTypes = {
  customFields: PropTypes.shape({
    channelPath: PropTypes.string.tag({
      label: 'Path',
      group: 'Channel',
      description:
        'Path to the feed, excluding the domain, defaults to /arcio/mrss',
      defaultValue: '/arcio/mrss/',
    }),
    selectVideo: PropTypes.kvp.tag({
      label: 'Select video using',
      group: 'Video',
      description:
        'This criteria is used to filter videos encoded in the streams array',
      defaultValue: {
        height: '',
        width: '',
        bitrate: 5400,
        stream_type: 'mp4',
      },
    }),
    ...generatePropsForFeed('rss', PropTypes),
  }),
}
Mrss.label = 'MRSS'
export default Consumer(Mrss)
