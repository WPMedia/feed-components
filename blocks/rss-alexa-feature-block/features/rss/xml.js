import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import get from 'lodash/get'
import moment from 'moment'
import getProperties from 'fusion:properties'
import { resizerKey } from 'fusion:environment'
import { BuildContent } from '@wpmedia/feeds-content-elements'
import { generatePropsForFeed } from '@wpmedia/feeds-prop-types'
import { buildResizerURL } from '@wpmedia/feeds-resizer'

const jmespath = require('jmespath')

const cheerio = require('cheerio')

const getURLType = (url) => {
  const imgRegex = new RegExp(/(jpe?g|png|webp|PNG|JPG|WEBP)$/);
  const audioRegex = new RegExp('^(https?|ftp|file):\/\/(www.)?(.*?)\.(mp3)$');
  const videoRegex = new RegExp('^(https?|ftp|file):\/\/(www.)?(.*?)\.(mp4|ts)$');
  const uri = url.split(".");
  const type = uri[uri.length -1];
  const defaultType =  {label: 'Audio',type: `audio/${type}`};
  if(imgRegex.test(url)) {
    return {label: 'image',type: `image/${type}`};
  }
  if(audioRegex.test(url)){
    return {label: 'Audio',type: `audio/${type}`};
  }
  if(videoRegex.test(url)){
    return {label: 'Video',type: `video/${type}`};
  }
  
  return defaultType;
}

const rssTemplate = (
  elements,
  {
    audioAvailable,
    channelTitle,
    channelDescription,
    channelCopyright,
    channelTTL,
    channelCategory,
    channelLogo,
    itemTitle,
    pubDate,
    itemCategory,
    includeContent,
    resizerURL,
    resizerWidth,
    resizerHeight,
    domain,
    feedTitle,
    feedLanguage,
    rssBuildContent,
  },
) => ({
  rss: {
    '@version': '2.0',
    channel: {
      ...(channelTitle && { title: channelTitle }),
      link: `${domain}`,
      ...(channelDescription && { description: channelDescription }),
      lastBuildDate: moment
        .utc(new Date())
        .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
      ...(feedLanguage && { language: feedLanguage }),
      ...(channelCategory && { category: channelCategory }),
      ...(channelCopyright && {
        copyright: channelCopyright,
      }), // TODO Add default logic
      ...(channelTTL && { ttl: channelTTL }),

      ...(channelLogo && {
        image: {
          url: buildResizerURL(channelLogo, resizerKey, resizerURL),
          title: channelTitle || feedTitle,
          link: domain,
        },
      }),

      item: elements.map((s) => {
        let body, category,
        enclosureurl = jmespath.search(s, audioAvailable);
        const url = `${domain}${s.website_url || s.canonical_url || ''}`
        return {
          title: jmespath.search(s, itemTitle) || '',
          link: url,
          guid: s._id,
          pubDate: s[pubDate],
          ...(enclosureurl  &&
          {
            enclosure: {
              '@url': enclosureurl,
              '@label':  getURLType(enclosureurl).label,
              '@group': getURLType(enclosureurl).label,
              '@type': getURLType(enclosureurl).type
            }
          }),
          ...(itemCategory &&
            (category = jmespath.search(s, itemCategory)) &&
            category && { category: category }),
            description: cheerio
            .load(
              rssBuildContent.parse(
                s.content_elements || [],
                includeContent,
                domain,
                resizerKey,
                resizerURL,
                resizerWidth,
                resizerHeight,
              ) || '',
            )
            .text()
        }
      }),
    },
  },
})

export function Rss({ globalContent, customFields, arcSite }) {
  const {
    resizerURL = '',
    feedDomainURL = '',
    feedTitle = '',
    feedLanguage = '',
  } = getProperties(arcSite)

  const { width = 0, height = 0 } = customFields.resizerKVP || {};
  

  const rssBuildContent = new BuildContent()
 
  // can't return null for xml return type, must return valid xml template
  return rssTemplate(get(globalContent, 'content_elements', []), {
    ...customFields,
    resizerURL,
    resizerWidth: width,
    resizerHeight: height,
    domain: feedDomainURL,
    feedTitle,
    feedLanguage,
    rssBuildContent,
  })
}

Rss.propTypes = {
  customFields: PropTypes.shape({
    audioAvailable: PropTypes.string.tag({
      label: 'Audio',
      group: 'Audio',
      description: 'description',
      defaultValue: `content_elements[?type=='audio'].streams[].url|[0]`,
    }),
    ...generatePropsForFeed('rss', PropTypes, ['featuredImage', 'includePromo']),
  }),
}
Rss.label = 'RSS Alexa'
export default Consumer(Rss)
