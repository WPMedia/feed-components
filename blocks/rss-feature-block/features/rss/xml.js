import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import get from 'lodash/get'
import moment from 'moment'
import getProperties from 'fusion:properties'
import { resizerKey } from 'fusion:environment'
import { BuildContent } from '@wpmedia/feeds-content-elements'
import { generatePropsForFeed } from '@wpmedia/feeds-prop-types'
import { buildResizerURL } from '@wpmedia/feeds-resizer'

const jmespath = require('jmespath')

const rssTemplate = (
  elements,
  {
    channelTitle,
    channelDescription,
    channelPath,
    channelCopyright,
    channelTTL,
    channelUpdatePeriod,
    channelUpdateFrequency,
    channelCategory,
    channelLogo,
    imageTitle,
    imageCaption,
    imageCredits,
    itemTitle,
    itemDescription,
    pubDate,
    itemCategory,
    includePromo,
    includeContent,
    resizerURL,
    domain,
    feedTitle,
    feedLanguage,
    rssBuildContent,
  },
) => ({
  rss: {
    '@xmlns:atom': 'http://www.w3.org/2005/Atom',
    '@xmlns:content': 'http://purl.org/rss/1.0/modules/content/',
    '@xmlns:dc': 'http://purl.org/dc/elements/1.1/',
    '@xmlns:sy': 'http://purl.org/rss/1.0/modules/syndication/',
    '@version': '2.0',
    ...(includePromo && {
      '@xmlns:media': 'http://search.yahoo.com/mrss/',
    }),
    channel: {
      title: `${channelTitle || feedTitle}`,
      link: `${domain}`,
      'atom:link': {
        '@href': `${domain}${channelPath}`,
        '@rel': 'self',
        '@type': 'application/rss+xml',
      },
      description: `${channelDescription || feedTitle + ' News Feed'}`,
      lastBuildDate: moment
        .utc(new Date())
        .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
      ...(feedLanguage && { language: feedLanguage }),
      ...(channelCategory && { category: channelCategory }),
      ...(channelCopyright && {
        copyright: channelCopyright,
      }), // TODO Add default logic
      ...(channelTTL && { ttl: channelTTL }),
      ...(channelUpdatePeriod && {
        'sy:updatePeriod': channelUpdatePeriod,
      }),
      ...(channelUpdateFrequency && {
        'sy:updateFrequency': channelUpdateFrequency,
      }),
      ...(channelLogo && {
        image: {
          url: buildResizerURL(channelLogo, resizerKey, resizerURL),
          title: `${channelTitle || feedTitle}`,
          link: `${domain}`,
        },
      }),

      item: elements.map((s) => {
        let author, body, category
        const url = `${domain}${s.website_url || s.canonical_url}`
        const img =
          s.promo_items && (s.promo_items.basic || s.promo_items.lead_art)
        return {
          title: `${jmespath.search(s, itemTitle)}`,
          link: url,
          guid: {
            '#': url,
            '@isPermaLink': true,
          },
          ...((author = jmespath.search(s, 'credits.by[].name')) &&
            author && {
              'dc:creator': author.join(','),
            }),
          description: { $: jmespath.search(s, itemDescription) },
          pubDate: moment
            .utc(s[pubDate])
            .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
          ...(itemCategory &&
            (category = jmespath.search(s, itemCategory)) &&
            category && { category: category }),
          ...(includeContent !== '0' &&
            (body = rssBuildContent.parse(
              s.content_elements,
              includeContent,
              domain,
              resizerKey,
              resizerURL,
            )) &&
            body && {
              'content:encoded': {
                $: body,
              },
            }),
          ...(includePromo &&
            img &&
            img.url && {
              'media:content': {
                '@type': 'image/jpeg',
                '@url': buildResizerURL(img.url, resizerKey, resizerURL),
                ...(jmespath.search(img, imageCaption) && {
                  'media:description': {
                    '@type': 'plain',
                    $: jmespath.search(img, imageCaption),
                  },
                }),
                ...(jmespath.search(img, imageTitle) && {
                  'media:title': {
                    $: jmespath.search(img, imageTitle),
                  },
                }),
                ...((jmespath.search(img, imageCredits) || []).length && {
                  'media:credit': {
                    '@role': 'author',
                    '@scheme': 'urn:ebu',
                    '#': jmespath.search(img, imageCredits).join(','),
                  },
                }),
              },
            }),
        }
      }),
    },
  },
})

export function Rss({ globalContent, customFields, arcSite }) {
  const {
    resizerURL = '',
    feedDomainURL = '',
    feedTitle = '',
    feedLanguage = '',
  } = getProperties(arcSite)

  const rssBuildContent = new BuildContent()

  // can't return null for xml return type, must return valid xml template
  return rssTemplate(get(globalContent, 'content_elements', []), {
    ...customFields,
    resizerURL,
    domain: feedDomainURL,
    feedTitle,
    feedLanguage,
    rssBuildContent,
  })
}

Rss.propTypes = {
  customFields: PropTypes.shape({
    channelPath: PropTypes.string.tag({
      label: 'Path',
      group: 'Channel',
      description:
        'Path to the feed, excluding the domain, defaults to /arcio/rss',
      defaultValue: '/arcio/rss/',
    }),
    ...generatePropsForFeed('rss', PropTypes),
  }),
}
Rss.label = 'Standard RSS'
export default Consumer(Rss)
