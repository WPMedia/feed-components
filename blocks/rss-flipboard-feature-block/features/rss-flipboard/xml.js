import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import get from 'lodash/get'
import moment from 'moment'
import getProperties from 'fusion:properties'
import { resizerKey } from 'fusion:environment'
import { BuildContent } from '@wpmedia/feeds-content-elements'
import { generatePropsForFeed } from '@wpmedia/feeds-prop-types'
import { buildResizerURL } from '@wpmedia/feeds-resizer'

const jmespath = require('jmespath')

const rssTemplate = (
  elements,
  {
    channelTitle,
    channelDescription,
    channelPath,
    channelCopyright,
    channelTTL,
    channelUpdatePeriod,
    channelUpdateFrequency,
    channelCategory,
    channelLogo,
    imageTitle,
    imageCaption,
    imageCredits,
    itemTitle,
    itemDescription,
    pubDate,
    itemCredits,
    itemCategory,
    includeContent,
    resizerURL,
    resizerWidth,
    resizerHeight,
    domain,
    feedTitle,
    feedLanguage,
    flipboardBuildContent,
  },
) => ({
  rss: {
    '@xmlns:atom': 'http://www.w3.org/2005/Atom',
    '@xmlns:content': 'http://purl.org/rss/1.0/modules/content/',
    '@xmlns:dc': 'http://purl.org/dc/elements/1.1/',
    '@xmlns:sy': 'http://purl.org/rss/1.0/modules/syndication/',
    '@xmlns:media': 'http://search.yahoo.com/mrss/',
    '@version': '2.0',
    channel: {
      title: { $: channelTitle || feedTitle },
      link: domain,
      'atom:link': {
        '@href': `${domain}${channelPath}`,
        '@rel': 'self',
        '@type': 'application/rss+xml',
      },
      description: { $: channelDescription || `${feedTitle} News Feed` },
      ...(feedLanguage && { language: feedLanguage }),
      lastBuildDate: moment
        .utc(new Date())
        .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
      ...(channelCategory && { category: channelCategory }),
      ...(channelCopyright && {
        copyright: channelCopyright,
      }), // TODO Add default logic
      ...(channelTTL && { ttl: channelTTL }),
      ...(channelUpdatePeriod && {
        'sy:updatePeriod': channelUpdatePeriod,
      }),
      ...(channelUpdateFrequency && {
        'sy:updateFrequency': channelUpdateFrequency,
      }),
      ...(channelLogo && {
        image: {
          url: buildResizerURL(channelLogo, resizerKey, resizerURL),
          title: channelTitle || feedTitle,
          link: domain,
        },
      }),

      item: elements.map((s) => {
        let author, body, category
        const url = `${domain}${s.website_url || s.canonical_url || ''}`
        const img =
          s.promo_items && (s.promo_items.basic || s.promo_items.lead_art)
        return {
          title: { $: jmespath.search(s, itemTitle) || '' },
          link: url,
          description: { $: jmespath.search(s, itemDescription) || '' },
          guid: {
            '#': url,
            '@isPermaLink': true,
          },
          ...((author = jmespath.search(s, itemCredits)) &&
            author && {
              'dc:creator': { $: author.join(', ') },
            }),
          pubDate: moment
            .utc(s[pubDate])
            .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
          ...(itemCategory &&
            (category = jmespath.search(s, itemCategory)) &&
            category && { category: category }),
          ...(includeContent !== 0 &&
            (body = flipboardBuildContent.parse(
              s.content_elements || [],
              includeContent,
              domain,
              resizerKey,
              resizerURL,
              resizerWidth,
              resizerHeight,
            )) &&
            body && {
              'content:encoded': {
                $: body,
              },
            }),
          ...(img &&
            img.url && {
              'media:content': {
                '@medium': 'image',
                '@type': 'image/jpeg',
                '@url': buildResizerURL(
                  img.url,
                  resizerKey,
                  resizerURL,
                  resizerWidth,
                  resizerHeight,
                ),
                ...(img.width && {
                  '@width': resizerWidth || img.width,
                }),
                ...(img.height && {
                  '@height': resizerHeight || img.height,
                }),
                ...(jmespath.search(img, imageCaption) && {
                  'media:description': {
                    '@type': 'plain',
                    $: jmespath.search(img, imageCaption),
                  },
                }),
                ...(jmespath.search(img, imageTitle) && {
                  'media:title': {
                    $: jmespath.search(img, imageTitle),
                  },
                }),
                ...((jmespath.search(img, imageCredits) || []).length && {
                  'media:credit': {
                    '@role': 'author',
                    '@scheme': 'urn:ebu',
                    '#': jmespath.search(img, imageCredits).join(','),
                  },
                }),
              },
            }),
        }
      }),
    },
  },
})

export function FlipboardRss({ globalContent, customFields, arcSite }) {
  const {
    resizerURL = '',
    feedDomainURL = '',
    feedTitle = '',
    feedLanguage = '',
  } = getProperties(arcSite)
  const { width = 0, height = 0 } = customFields.resizerKVP || {}

  function FlipboardBuildContent() {
    BuildContent.call(this)

    this.image = (
      element,
      resizerKey,
      resizerURL,
      resizerWidth,
      resizerHeight,
    ) => {
      return {
        figure: {
          img: {
            '@': {
              alt: element.caption || '',
              src: buildResizerURL(
                element.url,
                resizerKey,
                resizerURL,
                resizerWidth,
                resizerHeight,
              ),
              ...(element.width && { width: resizerWidth || element.width }),
              ...(element.height && {
                height: resizerHeight || element.height,
              }),
            },
          },
          ...(element.caption && { figcaption: element.caption }),
        },
      }
    }
  }

  const flipboardBuildContent = new FlipboardBuildContent()

  // can't return null for xml return type, must return valid xml template
  return rssTemplate(get(globalContent, 'content_elements', []), {
    ...customFields,
    resizerURL,
    resizerWidth: width,
    resizerHeight: height,
    domain: feedDomainURL,
    feedTitle,
    feedLanguage,
    flipboardBuildContent,
  })
}

FlipboardRss.propTypes = {
  customFields: PropTypes.shape({
    channelPath: PropTypes.string.tag({
      label: 'Path',
      group: 'Channel',
      description:
        'Path to the feed excluding the domain, defaults to /arc/outboundfeeds/flipboard',
      defaultValue: '/arc/outboundfeeds/flipboard/',
    }),
    ...generatePropsForFeed('rss', PropTypes, ['channelPath', 'includePromo']),
  }),
}

FlipboardRss.label = 'RSS Flipboard'
export default Consumer(FlipboardRss)
