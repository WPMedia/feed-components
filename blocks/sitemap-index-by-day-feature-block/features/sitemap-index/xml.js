import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import getProperties from 'fusion:properties'
import moment from 'moment'
import URL from 'url'

const sitemapIndexTemplate = ({
  maxDays,
  feedPath,
  feedParam,
  feedDates2Split,
  section,
  domain,
  buildIndexes,
}) => ({
  sitemapindex: {
    '@xmlns': 'http://www.sitemaps.org/schemas/sitemap/0.9',
    sitemap: buildIndexes(
      maxDays,
      domain,
      feedPath,
      section,
      feedParam,
      feedDates2Split,
    ),
  },
})

export function SitemapIndexByDay({
  globalContent,
  customFields,
  arcSite,
  requestUri,
}) {
  const { feedDomainURL = 'http://localhost' } = getProperties(arcSite)
  const pathList = new URL.URL(requestUri, feedDomainURL).pathname.split(
    customFields.feedName,
  )
  const section = pathList && pathList.length === 2 ? pathList[1] : ''

  // numberOfDays can be an int or a date
  let maxDays
  if (customFields.numberOfDays.match(/\d{4}-\d{2}-\d{2}/)) {
    const start = moment.utc(new Date()).startOf('day')
    const end = moment.utc(customFields.numberOfDays, 'YYYY-MM-DD', true)
    if (end.isValid() && end.isBefore(start) && end.isAfter('1994-12-31')) {
      maxDays = start.diff(end, 'days') + 1
    }
  }
  if (!maxDays) {
    maxDays =
      (!isNaN(customFields.numberOfDays) &&
        parseInt(customFields.numberOfDays)) ||
      2
  }

  const buildIndexes = (
    maxDays,
    feedDomainUrl,
    feedPath,
    section,
    feedParam,
    feedDates2Split,
  ) => {
    const arr = []
    const now = moment.utc(new Date())
    const splitAllDates =
      feedDates2Split.all || feedDates2Split.All || feedDates2Split.ALL
    const feedPathKeys = Object.keys(feedPath).map((i) => parseInt(i))
    let pathValue = '/arc/outboundfeeds/sitemap/'
    for (let i = 0; i < maxDays; i++) {
      const pathIndex = feedPathKeys.indexOf(i)
      if (pathIndex !== -1) pathValue = feedPath[i]
      const formattedDate = i === 0 ? 'latest' : now.format('YYYY-MM-DD')
      const numSplits = feedDates2Split[formattedDate] || splitAllDates
      if (numSplits) {
        for (let splits = 1; splits <= numSplits; splits++) {
          arr.push({
            loc: `${feedDomainURL}${pathValue}${section}${formattedDate}-${splits}/?${
              feedParam || ''
            }`,
          })
        }
      } else {
        arr.push({
          loc: `${feedDomainURL}${pathValue}${section}${formattedDate}/?${
            feedParam || ''
          }`,
        })
      }
      now.subtract(1, 'days')
    }
    return arr
  }

  // can't return null for xml return type, must return valid xml template
  return sitemapIndexTemplate({
    ...customFields,
    section,
    domain: feedDomainURL,
    maxDays,
    buildIndexes,
  })
}

SitemapIndexByDay.propTypes = {
  customFields: PropTypes.shape({
    numberOfDays: PropTypes.string.tag({
      label: 'Number of Days to include',
      group: 'Format',
      description:
        'How many days should be included in the Sitemap Index. Enter a date in YYYY-MM-DD format to use a specific end date. The earliest date is 1995-01-01.',
      defaultValue: '2',
    }),
    feedPath: PropTypes.kvp.tag({
      label: 'Number of days and sitemap path',
      group: 'Format',
      description:
        'Number of days and resolver to use.  Each higher resolver number has a longer TTL value to reduce Content-API load.',
      defaultValue: {
        0: '/arc/outboundfeeds/sitemap/',
        2: '/arc/outboundfeeds/sitemap2/',
        7: '/arc/outboundfeeds/sitemap3/',
      },
    }),
    feedName: PropTypes.string.tag({
      label: 'Sitemap-Index Name ',
      group: 'Format',
      description: 'Name of the sitemap-index feed in the URL',
      defaultValue: '/sitemap-index/',
    }),
    feedParam: PropTypes.string.tag({
      label: 'Additional URL Parameters',
      group: 'Format',
      description:
        'Optional parameters to append to URL. Separate values with an &, do not include a ?',
      defaultValue: 'outputType=xml',
    }),
    feedDates2Split: PropTypes.kvp.tag({
      label: 'Dates with large results',
      group: 'Format',
      description:
        'Dates that have more than 1000 results might time out. You can enter a date in YYYY-MM-DD format and the number of links of 1000 records to generate. Use All to do this on every day.',
      defaultValue: {},
    }),
  }),
}
SitemapIndexByDay.label = 'Sitemap Index By Day'
export default Consumer(SitemapIndexByDay)
