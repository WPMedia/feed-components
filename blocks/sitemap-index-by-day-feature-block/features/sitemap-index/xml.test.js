// eslint-disable-next-line no-unused-vars
import Consumer from 'fusion:consumer'
import { SitemapIndexByDay } from './xml'

jest.useFakeTimers('modern').setSystemTime(new Date('2021-04-09').getTime())

it('returns template with default values', () => {
  const sitemapindex = SitemapIndexByDay({
    arcSite: 'demo',
    customFields: {
      feedPath: {},
      feedName: '/sitemap-index/',
      feedParam: 'outputType=xml',
      numberOfDays: '2',
      feedDates2Split: {},
    },
    requestUri: '/sitemap-index/?outputType=xml',
  })
  expect(sitemapindex).toMatchSnapshot()
})

it('returns template with set end date', () => {
  const sitemapindex = SitemapIndexByDay({
    arcSite: 'demo',
    customFields: {
      feedPath: {
        0: '/arc/outboundfeeds/sitemap/',
        2: '/arc/outboundfeeds/sitemap2/',
        7: '/arc/outboundfeeds/sitemap3/',
      },
      feedName: '/sitemap-index/',
      feedParam: 'outputType=xml',
      numberOfDays: '2021-04-01',
      feedDates2Split: { '2021-04-06': '4' },
    },
    requestUri: '/sitemap-index/?outputType=xml',
  })
  expect(sitemapindex).toMatchSnapshot()
})
