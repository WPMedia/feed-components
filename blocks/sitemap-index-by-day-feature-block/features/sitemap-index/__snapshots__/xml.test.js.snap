// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid date format, should return 2 1`] = `
Object {
  "sitemapindex": Object {
    "@xmlns": "http://www.sitemaps.org/schemas/sitemap/0.9",
    "sitemap": Array [
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap-news/latest/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap-news/2021-04-08/?outputType=xml",
      },
    ],
  },
}
`;

exports[`invalid date, should return 2 1`] = `
Object {
  "sitemapindex": Object {
    "@xmlns": "http://www.sitemaps.org/schemas/sitemap/0.9",
    "sitemap": Array [
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap-news/latest/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap-news/2021-04-08/?outputType=xml",
      },
    ],
  },
}
`;

exports[`returns template with default values 1`] = `
Object {
  "sitemapindex": Object {
    "@xmlns": "http://www.sitemaps.org/schemas/sitemap/0.9",
    "sitemap": Array [
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap/latest/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap/2021-04-08/?outputType=xml",
      },
    ],
  },
}
`;

exports[`returns template with empty values 1`] = `
Object {
  "sitemapindex": Object {
    "@xmlns": "http://www.sitemaps.org/schemas/sitemap/0.9",
    "sitemap": Array [
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap/latest/",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap/2021-04-08/",
      },
    ],
  },
}
`;

exports[`returns template with set end date 2021-04-01 1`] = `
Object {
  "sitemapindex": Object {
    "@xmlns": "http://www.sitemaps.org/schemas/sitemap/0.9",
    "sitemap": Array [
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap/latest/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap/2021-04-08/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap2/2021-04-07/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap2/2021-04-06-1/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap2/2021-04-06-2/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap2/2021-04-06-3/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap2/2021-04-06-4/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap2/2021-04-05/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap2/2021-04-04/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap2/2021-04-03/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap3/2021-04-02/?outputType=xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/arc/outboundfeeds/sitemap3/2021-04-01/?outputType=xml",
      },
    ],
  },
}
`;

exports[`returns template without outputType 1`] = `
Object {
  "sitemapindex": Object {
    "@xmlns": "http://www.sitemaps.org/schemas/sitemap/0.9",
    "sitemap": Array [
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/sitemap/latest/",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/sitemap/2021-04-08/",
      },
    ],
  },
}
`;

exports[`sitemaps at root 1`] = `
Object {
  "sitemapindex": Object {
    "@xmlns": "http://www.sitemaps.org/schemas/sitemap/0.9",
    "sitemap": Array [
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/sitemap-latest.xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/sitemap-2021-04-08.xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/sitemap-2021-04-07.xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/sitemap-2021-04-06-1.xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/sitemap-2021-04-06-2.xml",
      },
      Object {
        "loc": "http://demo-prod.origin.arcpublishing.com/sitemap-2021-04-05.xml",
      },
    ],
  },
}
`;
